File: leetcode/go/645.set-mismatch.go
func findErrorNums(nums []int) []int {
    duplicate := 0
    for _, n := range nums {
        x := Abs(n)-1
        if nums[x] < 0 {
            duplicate = x + 1
        } else {
            nums[x] *= -1
        }
    }
    for i, n := range nums {
        if n > 0 {
            return []int{duplicate, i+1}
        }
    }
    return []int{0, 0}
}

func Abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}

File: leetcode/rust-leetcode/src/645.set-mismatch.rs
impl Solution {
    pub fn find_error_nums(mut nums: Vec<i32>) -> Vec<i32> {
        let mut duplicate = 0;
        for i in 0..nums.len() {
            let n = nums[i].abs() as usize - 1;
            if nums[n] < 0 {
                duplicate = n as i32 + 1;
            } else {
                nums[n] *= -1;
            }
        }
        for (i, &n) in nums.iter().enumerate() {
            if n > 0 {
                return vec![duplicate, i as i32 + 1]
            }
        }
        nums
    }
}

